korma/src/commonMain/kotlin/com/soywiz/korma/geom/Quaternion.kt

```diff
@@ -84,54 +89,71 @@ data class Quaternion(
+        }
     }
 
-    fun setToNlerp(left: Quaternion, right: Quaternion, t: Double): Quaternion {
-        val sign = if (dotProduct(left, right) < 0) -1 else +1
-        return setToFunc { (1f - t) * left[it] + t * right[it] * sign }.normalize()
+    fun setToNlerp(left: Quaternion, right: Quaternion, t: Float): Quaternion {
+        val sign = if (dotProduct(left, right) < 0f) -1f else +1f
+        setToFunc { (1f - t) * left[it] + t * right[it] * sign }
+        normalize()
+        return this
     }
 
-    fun setToInterpolated(left: Quaternion, right: Quaternion, t: Double): Quaternion = setToSlerp(left, right, t)
+    fun setToInterpolated(left: Quaternion, right: Quaternion, t: Float) = setToSlerp(left, right, t)
 
-    fun setFromRotationMatrix(m: Matrix3D) = this.apply {
-        val q = this
+    fun setFromRotationMatrix(m: Matrix3D): Unit = let { q ->
         m.apply {
             val t = v00 + v11 + v22
             when {
                 t > 0 -> {
-                    val s = 0.5 / sqrt(t + 1.0)
-                    q.setTo(((v21 - v12) * s), ((v02 - v20) * s), ((v10 - v01) * s), (0.25 / s))
+                    val s = 0.5f / sqrt(t + 1f)
+                    q.setTo(((v21 - v12) * s), ((v02 - v20) * s), ((v10 - v01) * s), (0.25f / s))
                 }
                 v00 > v11 && v00 > v22 -> {
-                    val s = 2.0 * sqrt(1.0 + v00 - v11 - v22)
-                    q.setTo((0.25 * s), ((v01 + v10) / s), ((v02 + v20) / s), ((v21 - v12) / s))
+                    val s = 2f * sqrt(1f + v00 - v11 - v22)
+                    q.setTo((0.25f * s), ((v01 + v10) / s), ((v02 + v20) / s), ((v21 - v12) / s))
                 }
                 v11 > v22 -> {
-                    val s = 2.0 * sqrt(1.0 + v11 - v00 - v22)
-                    q.setTo(((v01 + v10) / s), (0.25 * s), ((v12 + v21) / s), ((v02 - v20) / s))
+                    val s = 2f * sqrt(1f + v11 - v00 - v22)
+                    q.setTo(((v01 + v10) / s), (0.25f * s), ((v12 + v21) / s), ((v02 - v20) / s))
                 }
                 else -> {
-                    val s = 2.0 * sqrt(1.0 + v22 - v00 - v11)
+                    val s = 2f * sqrt(1f + v22 - v00 - v11)
                     q.setTo(((v02 + v20) / s), ((v12 + v21) / s), (0.25f * s), ((v10 - v01) / s))
                 }
             }
         }
     }
 
-    fun normalize(v: Quaternion = this): Quaternion {
-        val length = 1.0 / Vector3D.length(v.x, v.y, v.z, v.w)
-        return this.setTo(v.x / length, v.y / length, v.z / length, v.w / length)
+    fun normalize(v: Quaternion = this): Float {
+        val originalLength = length
+        setTo(v.x / originalLength, v.y / originalLength, v.z / originalLength, v.w / originalLength)
+        return originalLength
     }
-    fun toMatrix(out: Matrix3D = Matrix3D()): Matrix3D = out.multiply(
+
+    /**
+     * Normalize, except when the length is 0
+     */
+    fun normalizeSafely(v: Quaternion = this): Float {
```

Comment:
Why don't make `normalize` do this instead? Is this the typical expected behaviour? Maybe normalize can do this, and if you believe some people would want to have NaNs/Infinites here, we can also provide `normalizeUnsafe`